File: packages/sdk/src/provider.ts
/**
 * Abstract integration provider interface.
 *
 * Relayfile supports multiple integration providers (Nango, Composio, etc.)
 * for syncing external service data into the Relayfile filesystem.
 *
 * Each provider maps its own webhook/event format into Relayfile's canonical
 * path + semantics model.
 */

import type { RelayFileClient } from "./client.js";
import type {
  FileQueryItem,
  FilesystemEvent,
  FileSemantics,
  QueuedResponse
} from "./types.js";

// ---------------------------------------------------------------------------
// Common types
// ---------------------------------------------------------------------------

/** Normalized webhook input from any provider */
export interface WebhookInput {
  /** Provider name (e.g., "github", "slack", "zendesk") */
  provider: string;
  /** Object type / model (e.g., "tickets", "commits", "messages") */
  objectType: string;
  /** Unique object ID within the provider */
  objectId: string;
  /** Event type (e.g., "created", "updated", "deleted") */
  eventType: string;
  /** Raw payload data */
  payload: Record<string, unknown>;
  /** Optional relations to other objects */
  relations?: string[];
  /** Provider-specific metadata (connection IDs, user IDs, etc.) */
  metadata?: Record<string, string>;
}

/** Options for listing files from a specific provider */
export interface ListProviderFilesOptions {
  provider: string;
  objectType?: string;
  status?: string;
  limit?: number;
  signal?: AbortSignal;
}

/** Options for watching provider events */
export interface WatchProviderEventsOptions {
  provider: string;
  pollIntervalMs?: number;
  cursor?: string;
  signal?: AbortSignal;
}

// ---------------------------------------------------------------------------
// Provider path mapping
// ---------------------------------------------------------------------------

const DEFAULT_PATH_PREFIXES: Record<string, string> = {
  zendesk: "/zendesk",
  shopify: "/shopify",
  github: "/github",
  stripe: "/stripe",
  slack: "/slack",
  linear: "/linear",
  jira: "/jira",
  hubspot: "/hubspot",
  salesforce: "/salesforce",
  gmail: "/gmail",
  notion: "/notion",
  asana: "/asana",
  trello: "/trello",
  intercom: "/intercom",
  freshdesk: "/freshdesk",
  discord: "/discord",
  twilio: "/twilio",
};

export function computeCanonicalPath(
  provider: string,
  objectType: string,
  objectId: string
): string {
  const prefix = DEFAULT_PATH_PREFIXES[provider] ?? `/${provider}`;
  return `${prefix}/${objectType}/${objectId}.json`;
}

// ---------------------------------------------------------------------------
// Abstract provider
// ---------------------------------------------------------------------------

export abstract class IntegrationProvider {
  protected readonly client: RelayFileClient;
  abstract readonly name: string;

  constructor(client: RelayFileClient) {
    this.client = client;
  }

  /**
   * Ingest a webhook event from this provider into Relayfile.
   * Each provider implementation normalizes its event format to WebhookInput,
   * then writes it to the canonical path.
   */
  abstract ingestWebhook(
    workspaceId: string,
    rawInput: unknown,
    signal?: AbortSignal
  ): Promise<QueuedResponse>;

  /**
   * Query files from a specific provider.
   */
  async getProviderFiles(
    workspaceId: string,
    options: ListProviderFilesOptions
  ): Promise<FileQueryItem[]> {
    const prefix = DEFAULT_PATH_PREFIXES[options.provider] ?? `/${options.provider}`;
    const pathFilter = options.objectType
      ? `${prefix}/${options.objectType}/`
      : `${prefix}/`;

    const properties: Record<string, string> = {
      provider: options.provider,
    };
    if (options.objectType) {
      properties["provider.object_type"] = options.objectType;
    }
    if (options.status) {
      properties["provider.status"] = options.status;
    }

    const allItems: FileQueryItem[] = [];
    let cursor: string | undefined;
    for (;;) {
      const response = await this.client.queryFiles(workspaceId, {
        path: pathFilter,
        properties,
        cursor,
        limit: options.limit,
        signal: options.signal,
      });
      allItems.push(...response.items);
      if (
        !response.nextCursor ||
        (options.limit && allItems.length >= options.limit)
      ) {
        break;
      }
      cursor = response.nextCursor;
    }
    return options.limit ? allItems.slice(0, options.limit) : allItems;
  }

  /**
   * Watch for file events from a specific provider.
   */
  async *watchProviderEvents(
    workspaceId: string,
    options: WatchProviderEventsOptions
  ): AsyncGenerator<FilesystemEvent, void, unknown> {
    const pollIntervalMs = options.pollIntervalMs ?? 5000;
    let cursor = options.cursor;

    for (;;) {
      if (options.signal?.aborted) return;
      const response = await this.client.getEvents(workspaceId, {
        provider: options.provider,
        cursor,
        signal: options.signal,
      });
      for (const event of response.events) {
        yield event;
      }
      if (response.nextCursor) {
        cursor = response.nextCursor;
      }
      if (options.signal?.aborted) return;
      await new Promise<void>((resolve) => {
        const timer = setTimeout(() => {
          options.signal?.removeEventListener("abort", onAbort);
          resolve();
        }, pollIntervalMs);
        const onAbort = () => {
          clearTimeout(timer);
          resolve();
        };
        options.signal?.addEventListener("abort", onAbort, { once: true });
      });
    }
  }
}

File: sdk/relayfile-sdk-py/src/relayfile/provider.py
from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import Iterator
from dataclasses import dataclass
from threading import Event
from time import sleep
from typing import Any

from .client import RelayFileClient


DEFAULT_PATH_PREFIXES: dict[str, str] = {
    "zendesk": "/zendesk",
    "shopify": "/shopify",
    "github": "/github",
    "stripe": "/stripe",
    "slack": "/slack",
    "linear": "/linear",
    "jira": "/jira",
    "hubspot": "/hubspot",
    "salesforce": "/salesforce",
    "gmail": "/gmail",
    "notion": "/notion",
    "asana": "/asana",
    "trello": "/trello",
    "intercom": "/intercom",
    "freshdesk": "/freshdesk",
    "discord": "/discord",
    "twilio": "/twilio",
}


@dataclass
class WebhookInput:
    provider: str
    object_type: str
    object_id: str
    event_type: str
    payload: dict[str, Any]
    relations: list[str] | None = None
    metadata: dict[str, str] | None = None


@dataclass
class ListProviderFilesOptions:
    provider: str
    object_type: str | None = None
    status: str | None = None
    limit: int | None = None


@dataclass
class WatchProviderEventsOptions:
    provider: str
    poll_interval_seconds: float = 5.0
    cursor: str | None = None
    stop_event: Event | None = None


def compute_canonical_path(provider: str, object_type: str, object_id: str) -> str:
    prefix = DEFAULT_PATH_PREFIXES.get(provider, f"/{provider}")
    return f"{prefix}/{object_type}/{object_id}.json"


class IntegrationProvider(ABC):
    def __init__(self, client: RelayFileClient) -> None:
        self._client = client

    @property
    def client(self) -> RelayFileClient:
        return self._client

    @property
    @abstractmethod
    def name(self) -> str:
        raise NotImplementedError

    @abstractmethod
    def ingest_webhook(
        self,
        workspace_id: str,
        raw_input: Any,
    ) -> dict[str, Any]:
        raise NotImplementedError

    def get_provider_files(
        self,
        workspace_id: str,
        options: ListProviderFilesOptions,
    ) -> list[dict[str, Any]]:
        prefix = DEFAULT_PATH_PREFIXES.get(options.provider, f"/{options.provider}")
        path_filter = (
            f"{prefix}/{options.object_type}/"
            if options.object_type
            else f"{prefix}/"
        )

        properties: dict[str, str] = {"provider": options.provider}
        if options.object_type:
            properties["provider.object_type"] = options.object_type
        if options.status:
            properties["provider.status"] = options.status

        all_items: list[dict[str, Any]] = []
        cursor = None
        while True:
            result = self.client.query_files(
                workspace_id,
                path=path_filter,
                provider=options.provider,
                properties=properties,
                cursor=cursor,
                limit=options.limit,
            )
            all_items.extend(result.get("items", []))
            cursor = result.get("nextCursor")
            if not cursor or (options.limit and len(all_items) >= options.limit):
                break
        return all_items[: options.limit] if options.limit else all_items

    def watch_provider_events(
        self,
        workspace_id: str,
        options: WatchProviderEventsOptions,
    ) -> Iterator[dict[str, Any]]:
        cursor = options.cursor
        while True:
            if options.stop_event and options.stop_event.is_set():
                return

            response = self.client.get_events(
                workspace_id,
                provider=options.provider,
                cursor=cursor,
            )
            for event in response.get("events", []):
                yield event

            next_cursor = response.get("nextCursor")
            if next_cursor:
                cursor = next_cursor

            if options.stop_event:
                if options.stop_event.wait(options.poll_interval_seconds):
                    return
            else:
                sleep(options.poll_interval_seconds)
